### PR TITLE
Add support for service filtering during scanning

### PR DESCRIPTION
**Pull request by burnhamd**
Added scan filtering for IOS, and Android.
Pertains to Issue #22

### CODE REVIEW

**Comment by burnhamd** [comment]
Closing to fix some issues.

**Comment by twyatt** [comment]
> Closing to fix some issues.

@burnhamd great start on this feature. Would it also be possible to setup `Scanner` to use a builder lambda (similar to how `Peripheral` can be configured)?

Then the filters could be an option within the builder lambda.

**Comment by burnhamd** [comment]
@twyatt I certainly could, but I feel like that would be overkill?  
I suppose if we want to support all the scan options on all three systems it might make sense as there are many options.  But if we take the lowest common denominator approach, then it seems like only service UUIDs would be in the filter and iOS supports very little scan tweaks.  Just let me know what you think.

**Comment by twyatt** [comment]
I think long term the builder lambda would be nice so that Kable's API is consistent in how various components are created. As you mentioned, when platforms diverge in the options they support, I think the builder lambda is a slightly nicer API for library consumers. 🤷‍♂️

But going with the parameter approach for now would also be fine, then can later transition the API to lambda and deprecate the parameter.

**Comment by burnhamd** [comment]
Just tested with IOS and Android.  I haven't changed to builder.  We are using this in production project, so I wanted to at least put it here in case it was desired.  Im all for the builder approach but just need to work on some other stuff first.

**Comment by twyatt** [comment]
> Just tested with IOS and Android. I haven't changed to builder. We are using this in production project, so I wanted to at least put it here in case it was desired. Im all for the builder approach but just need to work on some other stuff first.

Sounds good. Thanks for testing it out! Happy to get it in with the current (parameter) approach. I'll review the PR soon.

**Comment by twyatt** [comment]
@burnhamd you can ignore build failure. Apparently I need to update our build scripts to skip a step for forked PRs.

File: core/src/androidMain/kotlin/Scanner.kt [hunk 1]
```diff
@@ -47,7 +51,16 @@ public class AndroidScanner internal constructor() : Scanner {
                 cancel("Bluetooth scan failed", ScanFailedException(errorCode))
             }
         }
-        bluetoothAdapter.bluetoothLeScanner.startScan(callback)
+
+        val scanFilter =
+            filterServices?.map { ScanFilter.Builder().setServiceUuid(ParcelUuid(it)).build() }
+                ?.toMutableList()
```
Comment (twyatt) [inline core/src/androidMain/kotlin/Scanner.kt:57]: Minor formatting update and use `toList`:

```suggestion
    val scanFilter =
        filterServices
            ?.map { ScanFilter.Builder().setServiceUuid(ParcelUuid(it)).build() }
            ?.toList()
```

File: core/src/androidMain/kotlin/Scanner.kt [hunk 2]
```diff
@@ -47,7 +51,16 @@ public class AndroidScanner internal constructor() : Scanner {
...
+        bluetoothAdapter.bluetoothLeScanner.startScan(
+            scanFilter,
+            ScanSettings.Builder().build(),
+            callback
```
Comment (twyatt) [inline core/src/androidMain/kotlin/Scanner.kt:61]: nitpick: Trailing comma, please.

```suggestion
            callback,
```

File: core/src/appleMain/kotlin/CentralManager.kt [hunk 3]
```diff
@@ -29,7 +29,10 @@ public class CentralManager internal constructor() {
         options: Map<Any?, *>?,
     ) {
         withContext(dispatcher) {
-            cbCentralManager.scanForPeripheralsWithServices(services, options)
+            cbCentralManager.scanForPeripheralsWithServices(
+                serviceUUIDs = services?.map { it.toCBUUID() },
+                options = options
```
Comment (twyatt) [inline core/src/appleMain/kotlin/CentralManager.kt:34]: nitpick: Trailing comma, please.

```suggestion
                options = options,
```

File: core/src/appleMain/kotlin/Scanner.kt [hunk 4]
```diff
@@ -13,7 +13,10 @@ import platform.CoreBluetooth.CBManagerStatePoweredOn
 import platform.CoreBluetooth.CBManagerStateUnauthorized
 import platform.CoreBluetooth.CBManagerStateUnsupported
 
-public actual fun Scanner(): Scanner = AppleScanner(CentralManager.Default, services = null)
+public actual fun Scanner(services: List<Uuid>?): Scanner = AppleScanner(
+    central = CentralManager.Default,
+    services = services
```
Comment (twyatt) [inline core/src/appleMain/kotlin/Scanner.kt:18]: nitpick

```suggestion
    services = services,
```

File: core/src/commonMain/kotlin/Scanner.kt [hunk 5]
```diff
@@ -8,4 +9,4 @@ public interface Scanner {
 
 // todo: Add support for specifying services to filter against (for platforms that support it).
 // https://github.com/JuulLabs/kable/issues/22
```
Comment (twyatt) [inline core/src/commonMain/kotlin/Scanner.kt:11]: I think this PR covers #22 enough to close it out, so we can drop this comment.

```suggestion
```

File: core/src/jsMain/kotlin/Scanner.kt [hunk 6]
```diff
@@ -11,7 +12,14 @@ import org.w3c.dom.events.Event
...
 
-public actual fun Scanner(): Scanner = JsScanner(bluetooth, Options())
+public actual fun Scanner(services: List<Uuid>?): Scanner {
+    // Below is a bit hacky
+    // TODO make more general and have more options
+    val serviceArray = services?.map { it.toString() }?.toTypedArray()
+    val filter = if (serviceArray != null) Options.Filter.Services(serviceArray) as Options.Filter else null
+    val filterArray = if (filter != null) arrayOf(filter) else null
+    return JsScanner(bluetooth = bluetooth, options = Options(filters = filterArray))
+}
```
Comment (twyatt) [inline core/src/jsMain/kotlin/Scanner.kt:22]:
```suggestion
public actual fun Scanner(services: List<Uuid>?): Scanner {
    val filters = services
        ?.map { it.toString() }
        ?.toTypedArray()
        ?.let { arrayOf(Options.Filter.Services(it) as Options.Filter) }
    return JsScanner(
        bluetooth = bluetooth,
        options = Options(filters = filters),
    )
}
```

**Review by twyatt — APPROVED** [review]
Looks great! Thanks so much!

**Review by burnhamd — COMMENTED** [review]
All changes look great!

**Comment by twyatt** [comment]
> All changes look great!

@burnhamd did you want to make the requested changes to your `main` branch; or would you like me to make the changes and merge manually?

**Review by davidtaylor-juul — APPROVED** [review]
LGTM, +1 for @twyatt's suggestions.

**Comment by burnhamd** [comment]
> > All changes look great!
> 
> @burnhamd did you want to make the requested changes to your [`main`](https://github.com/burnhamd/kable) branch; or would you like me to make the changes and merge manually?

Sorry for delay.  Committed to my repo.  Should be good to go now.

File: core/src/androidMain/kotlin/Scanner.kt [hunk 7]
```diff
@@ -47,7 +51,17 @@ public class AndroidScanner internal constructor() : Scanner {
                 cancel("Bluetooth scan failed", ScanFailedException(errorCode))
             }
         }
-        bluetoothAdapter.bluetoothLeScanner.startScan(callback)
+
+    val scanFilter =
+        filterServices
+            ?.map { ScanFilter.Builder().setServiceUuid(ParcelUuid(it)).build() }
+            ?.toList()
```
Comment (twyatt) [inline core/src/androidMain/kotlin/Scanner.kt:58]: Wups, when I suggested this change, I think I suggested the wrong indent.

```suggestion
        val scanFilter =
            filterServices
                ?.map { ScanFilter.Builder().setServiceUuid(ParcelUuid(it)).build() }
                ?.toList()
```

**Review by twyatt — APPROVED** [review]
Awesome job on this! Thanks so much for the contribution!

**Pull request merged**